src/main/java/ch/njol/skript/variables/HintManager.java

```diff
@@ -0,0 +1,468 @@
+ * # thus, hint for {_x} is Integer.class
+ * }
+ * </pre>
+ * @see ParserInstance#getHintManager()
+ */
+@ApiStatus.Experimental
+public class HintManager {
+
+	private record Scope(Map<String, Set<Class<?>>> hintMap, boolean isSection) { }
+
+	private final LinkedList<Scope> typeHints = new LinkedList<>();
+	private boolean isActive;
+
+	public HintManager(boolean active) {
+		isActive = active;
+	}
+
+	/**
+	 * Marks this hint manager as active or inactive.
+	 * An inactive hint manager does not collect hints.
+	 * That is, actions such as setting, adding, etc. have no effect on the currently stored hints.
+	 * Additionally, any calls to obtain hints will always result in an empty collection.
+	 * As a result, type hints are effectively not used.
+	 * @param active Whether this hint manager should be active.
+	 * @see #isActive
+	 */
+	public void setActive(boolean active) {
+		isActive = active;
+	}
+
+	/**
+	 * @return Whether this manager is active.
+	 * @see #setActive(boolean)
+	 */
+	public boolean isActive() {
+		ParserInstance parser = ParserInstance.get();
+		return isActive && parser.isActive() && parser.hasExperiment(Feature.TYPE_HINTS);
+	}
+
+	/**
+	 * Enters a new scope for storing hints.
+	 * Hints from the previous (current top-level) scope are copied over.
+	 * @param isSection Whether this scope represents a section in a trigger.
+	 * @see #exitScope()
+	 */
+	public void enterScope(boolean isSection) {
+		typeHints.push(new Scope(new HashMap<>(), isSection));
+		if (typeHints.size() > 1) { // copy existing values into new scope
+			mergeScope(1, 0, false);
+		}
+	}
+
+	/**
+	 * Exits the current (top-level) scope.
+	 * Hints from the exited scope will be copied over to the new top-level scope.
+	 * @see #enterScope(boolean)
+	 */
+	public void exitScope() {
+		if (typeHints.size() > 1) { // copy over updated hints
+			mergeScope(0, 1, false);
+		}
+		typeHints.pop();
+	}
+
+	/**
+	 * Resets (clears) all type hints for the current (top-level) scope.
+	 * Scopes are represented as integers, where {@code 0} represents the most recently entered scope
+	 * (i.e. the scope pushed by the most recent {@link #enterScope(boolean)} call).
+	 * @param sectionOnly Whether only scopes representing sections should be considered.
+	 */
+	public void clearScope(int level, boolean sectionOnly) {
+		if (!sectionOnly) {
+			typeHints.get(level).hintMap().clear();
+			return;
+		}
+		int currentLevel = 0;
+		var iterator = typeHints.iterator();
+		while (iterator.hasNext()) {
+			Scope scope = iterator.next();
+			if (!scope.isSection()) {
+				continue;
+			}
+			if (currentLevel == level) {
+				iterator.remove();
+				break;
+			}
+			currentLevel++;
+		}
+	}
+
+	/**
+	 * Copies hints from one scope to another.
+	 * Scopes are represented as integers, where {@code 0} represents the most recently entered scope
+	 * (i.e. the scope pushed by the most recent {@link #enterScope(boolean)} call).
```

Comment:
this should also explain what int is used for the second most recent scope to establish what the pattern is, as it's not really clear on first read